LogSumExp seems to be missing in SoftmaxCrossEntropyLoss (sparseLabel=false)

## Description
On latest MasterBranch (commit 4b516196) in SoftmaxCrossEntropyLoss (sparseLabel=false) line 85 
```
            loss = pred.mul(lab).neg().sum(new int[] {classAxis}, true);
```
looks like the LogSumExp term (red) is missing.
![grafik](https://user-images.githubusercontent.com/9622808/104430974-a90bc980-5587-11eb-8794-e80566641c1d.png)

## Proposed correction
```
            int[] axes = new int[] {classAxis};
            NDArray max = pred.max(axes, true);
            NDArray logSumExp = max.add((pred.sub(max)).exp().sum(axes, true).log());
            loss = logSumExp.sub(pred.mul(lab).sum(axes, true));
```

Hi @roywei,
the pull request "fix softmax flag" makes the behaviour clear. Now, I understand the approach :-) Thanks a lot.

To be solidly sure, that I get the correct gradients from the loss function I compared the gradients "handcalculated" against "autograd", once for tensorflow and once for DJL on mxnet (see below).
For tensorflow the gradients are the same - on DJL they seam to differ by a factor two. 
Please, check if you could see the cause - hopefully something wrong in my test ... then sorry for the noise.

**Tensorflow**
```
import tensorflow as tf

logits = [[-8.0, 10.0,  3.0],  [1.0, 2.0, 3.0]]
target = [[0.3, 0.3, 0.4], [0.1, 0.1, 0.8]]

probabilities = tf.nn.softmax(
    logits, axis=None, name=None
)
print("probabilities: ")
print(probabilities)

handcalculatedGradient = probabilities - target
print("handcalculated gradient (at logits): ")
print(handcalculatedGradient)

o = tf.Variable(logits)
with tf.GradientTape() as tape:
    loss = tf.nn.softmax_cross_entropy_with_logits(labels=target, logits=o)
print("loss: ")
print(loss)

dl_do = tape.gradient(loss, o)
print("autodiff gradient (at logits): ")
print(dl_do)

print("are handcalculated gradient and autodiff gradient equal?")
print(tf.math.equal(handcalculatedGradient , dl_do, name=None))
```
give
```
probabilities: 
tf.Tensor(
[[1.5216104e-08 9.9908900e-01 9.1105123e-04]
 [9.0030573e-02 2.4472848e-01 6.6524094e-01]], shape=(2, 3), dtype=float32)
handcalculated gradient (at logits): 
tf.Tensor(
[[-0.29999998  0.699089   -0.39908895]
 [-0.00996943  0.14472848 -0.13475907]], shape=(2, 3), dtype=float32)
loss: 
tf.Tensor([8.2009115  0.70760596], shape=(2,), dtype=float32)
autodiff gradient (at logits): 
tf.Tensor(
[[-0.29999998  0.699089   -0.39908895]
 [-0.00996943  0.14472848 -0.13475907]], shape=(2, 3), dtype=float32)
are handcalculated gradient and autodiff gradient equal?
tf.Tensor(
[[ True  True  True]
 [ True  True  True]], shape=(2, 3), dtype=bool)
```

**DJL on mxnet**
```
    @Test
    public void softmaxCrossEntropyNonSparseLabelsTest() {
        boolean fromLogit = true;
        boolean sparseLabel = false;

        try (NDManager manager = NDManager.newBaseManager()) {

            NDArray logits = manager.create(new float[] {-8.0f, 10.0f, 3.0f, 1.0f, 2.0f, 3.0f}, new Shape(2,3));
            NDArray label = manager.create(new float[] {0.3f, 0.3f, 0.4f, 0.1f, 0.1f, 0.8f}, new Shape(2,3));
            NDArray expectedLoss = manager.create(new float[] {8.2009115f,  0.70760596f}).mean();

            logits.attachGradient();
            try (GradientCollector gc = manager.getEngine().newGradientCollector()) {
                NDArray probabilities = logits.softmax(1);
                System.out.println("probabilities: ");
                System.out.println(probabilities);

                NDArray handcalculatedGradient = probabilities.sub(label);
                System.out.println("handcalculatedGradient: ");
                System.out.println(handcalculatedGradient);


                NDArray loss = new SoftmaxCrossEntropyLoss("SoftmaxCrossEntropyLoss", 1, 1, sparseLabel, !fromLogit).evaluate(new NDList(label), new NDList(logits));
                Assertions.assertAlmostEquals(loss, expectedLoss);
                gc.backward(loss);
                NDArray autograd = logits.getGradient();
                System.out.println("autograd: ");
                System.out.println(autograd);
                // The following assertion fails
                Assertions.assertAlmostEquals(autograd, handcalculatedGradient);
                // the assertion passes but the factor 2 is wrong
                // Assertions.assertAlmostEquals(handcalculatedGradient, autograd.mul(2));
            }
        }
    }
```
gives
```
probabilities: 
ND: (2, 3) gpu(0) float32
[[ 1.52161022e-08,  9.99088883e-01,  9.11051058e-04],
 [ 9.00305659e-02,  2.44728461e-01,  6.65240884e-01],
]

handcalculatedGradient: 
ND: (2, 3) gpu(0) float32
[[-0.3   ,  0.6991, -0.3991],
 [-0.01  ,  0.1447, -0.1348],
]

autograd: 
ND: (2, 3) gpu(0) float32
[[-0.15  ,  0.3495, -0.1995],
 [-0.005 ,  0.0724, -0.0674],
]


java.lang.AssertionError: 
Expected: ND: (2, 3) gpu(0) float32
[[-0.3   ,  0.6991, -0.3991],
 [-0.01  ,  0.1447, -0.1348],
]

Actual: ND: (2, 3) gpu(0) float32
[[-0.15  ,  0.3495, -0.1995],
 [-0.005 ,  0.0724, -0.0674],
]
```

--- a/api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java
+++ b/api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java
@@ -52,9 +52,10 @@ public class SoftmaxCrossEntropyLoss extends Loss {
      * @param name the name of the loss
      * @param weight the weight to apply on the loss value, default 1
      * @param classAxis the axis that represents the class probabilities, default -1
-     * @param sparseLabel whether labels are integer array or probabilities, default true
-     * @param fromLogit whether predictions are log probabilities or un-normalized numbers, default
-     *     false
+     * @param sparseLabel whether labels are 1-D integer array or 2-D probabilities of [batch_size,
+     *     n-class], default true
+     * @param fromLogit whether predictions are un-normalized numbers or log probabilities, if true,
+     *     logSoftmax will be applied to input, default true
      */
     public SoftmaxCrossEntropyLoss(
             String name, float weight, int classAxis, boolean sparseLabel, boolean fromLogit) {
@@ -69,7 +70,7 @@ public class SoftmaxCrossEntropyLoss extends Loss {
     @Override
     public NDArray evaluate(NDList label, NDList prediction) {
         NDArray pred = prediction.singletonOrThrow();
-        if (!fromLogit) {
+        if (fromLogit) {
             pred = pred.logSoftmax(classAxis);
         }
         NDArray loss;

--- a/api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java
+++ b/api/src/main/java/ai/djl/training/loss/SoftmaxCrossEntropyLoss.java
@@ -43,7 +43,7 @@ public class SoftmaxCrossEntropyLoss extends Loss {
      * @param name the name of the loss
      */
     public SoftmaxCrossEntropyLoss(String name) {
-        this(name, 1, -1, true, false);
+        this(name, 1, -1, true, true);
     }
 
     /**

--- a/integration/src/main/java/ai/djl/integration/tests/nn/BlockCoreTest.java
+++ b/integration/src/main/java/ai/djl/integration/tests/nn/BlockCoreTest.java
@@ -473,7 +473,7 @@ public class BlockCoreTest {
                     Assertions.assertAlmostEquals(result.size(), 2);
                     NDArray lossValue =
                             loss.evaluate(new NDList(labels), new NDList(result.head()));
-                    Assertions.assertAlmostEquals(lossValue.getFloat(), -18);
+                    Assertions.assertAlmostEquals(lossValue.getFloat(), 24.9533);
                     testEncode(manager, block);
                 }
             }
@@ -521,7 +521,9 @@ public class BlockCoreTest {
                     Assertions.assertAlmostEquals(result.size(), 2);
                     NDArray lossValue =
                             loss.evaluate(new NDList(labels), new NDList(result.head()));
-                    Assertions.assertAlmostEquals(lossValue.getFloat(), -908);
+                    // loss should be the same as testRNNTanh because outputs are equal for each
+                    // class
+                    Assertions.assertAlmostEquals(lossValue.getFloat(), 24.9533);
                     testEncode(manager, block);
                 }
             }
@@ -571,7 +573,7 @@ public class BlockCoreTest {
                     Assertions.assertAlmostEquals(result.size(), 3);
                     NDArray lossValue =
                             loss.evaluate(new NDList(labels), new NDList(result.head()));
-                    Assertions.assertAlmostEquals(lossValue.getFloat(), -16.340019);
+                    Assertions.assertAlmostEquals(lossValue.getFloat(), 24.9533);
                     testEncode(manager, block);
                 }
             }
@@ -628,7 +630,7 @@ public class BlockCoreTest {
                     Assertions.assertAlmostEquals(result.size(), 1);
                     NDArray lossValue =
                             loss.evaluate(new NDList(labels), new NDList(result.head()));
-                    Assertions.assertAlmostEquals(lossValue.getFloat(), -8.17537307E-4);
+                    Assertions.assertAlmostEquals(lossValue.getFloat(), 24.9533);
                     testEncode(manager, block);
                 }
             }

--- a/integration/src/main/java/ai/djl/integration/tests/training/LossTest.java
+++ b/integration/src/main/java/ai/djl/integration/tests/training/LossTest.java
@@ -46,11 +46,31 @@ public class LossTest {
     @Test
     public void softmaxCrossEntropyTest() {
         try (NDManager manager = NDManager.newBaseManager()) {
+            // test fromLogits=true, sparseLabel=true
             NDArray pred = manager.create(new float[] {1, 2, 3, 4, 5});
             NDArray label = manager.ones(new Shape(1));
             Assertions.assertAlmostEquals(
-                    Loss.softmaxCrossEntropyLoss().evaluate(new NDList(label), new NDList(pred)),
+                    Loss.softmaxCrossEntropyLoss("loss", 1, -1, true, true)
+                            .evaluate(new NDList(label), new NDList(pred)),
                     manager.create(3.45191431f));
+
+            // test fromLogits=false, sparseLabel=true
+            pred =
+                    manager.create(
+                            new float[] {4.0f, 2.0f, 1.0f, 0.0f, 5.0f, 1.0f}, new Shape(2, 3));
+            label = manager.create(new float[] {0, 1}, new Shape(2));
+            NDArray nonSparseLabel =
+                    manager.create(new float[] {1f, 0f, 0f, 0f, 1f, 0f}, new Shape(2, 3));
+            NDArray sparseOutput =
+                    Loss.softmaxCrossEntropyLoss()
+                            .evaluate(new NDList(label), new NDList(pred.logSoftmax(-1)));
+            // test fromLogits=false, sparseLabel=false
+            NDArray nonSparseOutput =
+                    Loss.softmaxCrossEntropyLoss("loss", 1, -1, false, false)
+                            .evaluate(new NDList(nonSparseLabel), new NDList(pred.logSoftmax(-1)));
+
+            Assertions.assertAlmostEquals(sparseOutput, nonSparseOutput);
+            Assertions.assertAlmostEquals(sparseOutput, manager.create(0.09729549f));
         }
     }
 

